java/ql/lib/semmle/code/java/frameworks/apache/Collections.qll

```diff
@@ -182,6 +204,695 @@ private class ApacheKeyValueModel extends SummaryModelCsv {
+          ".multimap;HashSetValuedHashMap;true;HashSetValuedHashMap;(MultiValuedMap);;Element of MapValue of Argument[0];Element of MapValue of Argument[-1];value",
+          ".multimap;HashSetValuedHashMap;true;HashSetValuedHashMap;(Map);;MapKey of Argument[0];MapKey of Argument[-1];value",
+          ".multimap;HashSetValuedHashMap;true;HashSetValuedHashMap;(Map);;MapValue of Argument[0];Element of MapValue of Argument[-1];value",
+          ".multimap;TransformedMultiValuedMap;true;transformingMap;;;MapKey of Argument[0];MapKey of ReturnValue;value",
+          ".multimap;TransformedMultiValuedMap;true;transformingMap;;;Element of MapValue of Argument[0];Element of MapValue of ReturnValue;value",
+          ".multimap;UnmodifiableMultiValuedMap;true;unmodifiableMultiValuedMap;(MultiValuedMap);;MapKey of Argument[0];MapKey of Argument[-1];value",
+          ".multimap;UnmodifiableMultiValuedMap;true;unmodifiableMultiValuedMap;(MultiValuedMap);;Element of MapValue of Argument[0];Element of MapValue of Argument[-1];value"
+        ]
+  }
+}
+
+/**
+ * Value-propagating models for the package `org.apache.commons.collections4.multiset`.
+ */
+private class ApacheMultiSetModel extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      ["org.apache.commons.collections4", "org.apache.commons.collections"] +
+        [
+          ".multiset;HashMultiSet;true;HashMultiSet;;;Element of Argument[0];Element of Argument[-1];value",
+          ".multiset;PredicatedMultiSet;true;predicatedMultiSet;;;Element of Argument[0];Element of ReturnValue;value",
+          ".multiset;SynchronizedMultiSet;true;synchronizedMultiSet;;;Element of Argument[0];Element of ReturnValue;value",
+          ".multiset;UnmodifiableMultiSet;true;unmodifiableMultiSet;;;Element of Argument[0];Element of ReturnValue;value"
+        ]
+  }
+}
+
+/**
+ * Value-propagating models for the package `org.apache.commons.collections4.properties`.
+ */
+private class ApachePropertiesModel extends SummaryModelCsv {
+  override predicate row(string row) {
+    row =
+      ["org.apache.commons.collections4", "org.apache.commons.collections"] +
+        [
+          ".properties;AbstractPropertiesFactory;true;load;(ClassLoader,String);;Argument[1];ReturnValue;taint",
+          ".properties;AbstractPropertiesFactory;true;load;(File);;Argument[0];ReturnValue;taint",
+          ".properties;AbstractPropertiesFactory;true;load;(InputStream);;Argument[0];ReturnValue;taint",
+          ".properties;AbstractPropertiesFactory;true;load;(Path);;Argument[0];ReturnValue;taint",
+          ".properties;AbstractPropertiesFactory;true;load;(Reader);;Argument[0];ReturnValue;taint",
+          ".properties;AbstractPropertiesFactory;true;load;(String);;Argument[0];ReturnValue;taint",
+          ".properties;AbstractPropertiesFactory;true;load;(URI);;Argument[0];ReturnValue;taint",
+          ".properties;AbstractPropertiesFactory;true;load;(URL);;Argument[0];ReturnValue;taint"
```

Comment:
Do we need to model `SortedProperties`. I couldn't find (yet) where we model `Properties` to account for that

Reply:
What would you model in `SortedProperties`? I don't think we model `Properties` for data flow at the moment. We probably should. We do model `Dictionary`, which is a superclass of it.